pymatgen/io/cp2k/inputs.py

```diff
@@ -0,0 +1,1364 @@
+"""
+This module defines the building blocks of a CP2K input file. The cp2k input structure is essentially a collection
+of "sections" which are similar to dictionary objects that activate modules of the cp2k executable, and then
+"keywords" which adjust variables inside of those modules. For example, FORCE_EVAL section will activate CP2K's ability
+to calculate forces, and inside FORCE_EVAL, the Keyword "METHOD can be set to "QS" to set the method of force evaluation
+to be the quickstep (DFT) module.
+
+A quick overview of the module:
+
+-- Section class defines the basis of Cp2k input and contains methods for manipulating these objects similarly to Dicts.
+-- Keyword class defines the keywords used inside of Section objects that changes variables in Cp2k program
+-- Cp2kInput class is special instantiation of Section that is used to represent the full cp2k calculation input.
+-- The rest of the classes are children of Section intended to make initialization of common sections easier.
+"""
+
+import os
+import copy
+import textwrap
+from monty.json import MSONable
+from monty.io import zopen
+from pymatgen.io.cp2k.utils import _postprocessor
+
+__author__ = "Nicholas Winner"
+__version__ = "0.2"
+__email__ = "nwinner@berkeley.edu"
+__date__ = "January 2019"
+
+
+class Section(MSONable):
+
+    """
+    Basic input representation of input to Cp2k. Activates functionality inside of the Cp2k executable.
+    """
+
+    # TODO: Not sure if I want to use this required_sections/keywords idea
+    required_sections: list = []
+    required_keywords: list = []
+    subsections: dict = {}  # Must exist before __init__ is even called
```

Comment:
So on the point of "required_*". It was an idea that I never properly implemented. It was inspired by how fireworks has required arguments as a way to require that certain sections have things in them. For example if you instantiate a "KIND" section to describe an atom type, you're required to say what pseudopotential you want to use. I didn't get around to implementing it because going through the manual to find what was required was some busy work I couldn't do at the time. I guess I still don't know if I want to use it. Thoughts?

As for `subsections` shadowing `self.subsections`. There was a problem I was having early on where subsections needed to exist as a variable with un-initiated versions of the Section class... I'll try to hunt down why that was so that it doesn't need to be written this way.

Reply:
It might have been an issue with `None`. We normally do something like 
``` python
def __init__(self, ... , subsections =None):
    self.subsections = subsections if subsections is not None else {}
`